### ISSUE TITLE
dbt.deferConfigPerProject doesn't work correctly with the "Execute DBT SQL" or "Compiled DBT Preview"

### ISSUE THREAD

**Issue by mostafa-k-m** [AltimateAI/vscode-dbt-power-user#1365]
### Expected behavior

After adding dbt.deferConfigPerProject config, My models should defer to upstream tables when I use the DBT Power User Ad hoc query, compiled preview or preview query results

### Actual behavior

Defer only works correctly when Build DBT model is used.

Defer Does not work with Ad hoc queries, compiled preview or preview query results

### Steps To Reproduce

follow the steps in this guide
https://docs.myaltimate.com/test/defertoprod/

use the local mode
under local path, you should choose the path containing the manifest file pointing to production tables.

your vscode settings should look like this:
```
    "dbt.deferConfigPerProject": {
        
        "<dbt project name>": {
            "deferToProduction": true,
            "manifestPathForDeferral": "<local path to target/live>",
            "manifestPathType": "local",
            "favorState": true
        }
    }
```

then:
- CMD+P > dbt Power User: Run dbt Model:  works as expected
- CMD+P > dbt Power User: compiled dbt preview: doesn't defer to production models as expected
- CMD+P > dbt Power User: Execute DBT SQL: doesn't defer to production models as expected

### Operating System

macOS 14.5 23F79 arm64

### dbt version

1.8.0

### dbt Adapter

databricks=1.8.3

### dbt Power User version

v0.43.6

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

**Comment by rafaelkrysciak** [comment]
I have the same behavior with dbt Power User version 0.50.5. I'm using dbt-core version 1.8.9 and connector teradata=1.8.2.

**Comment by cheyuriy** [comment]
Encountered similar issue, but was able to investigate it a bit further.

tl;dr "Compiled dbt Preview" feature breaks defer.

Scenario:
1. Open DBT project in VSCode. Defe is set up properly.
2. Open a model with a references to other models, which are present in "production" but is not present in your "dev" dataset (in terms of BigQuery)
3. Run "Build Model" - everything is fine, defer works
4. Run "Execute dbt SQL" to view Query Results - everything is fine, defer works
5. Run "Compiled dbt Preview" to view resulting SQL in a separate tab - the code is wrong, includes "dev" dataset, defer doesn't work
6. Try again to run "Execute dbt SQL" - error complaining that some tables are not found in "dev" dataset, defer doesn't work
7. Try again to run "Build Model" - everything is fine

To make everything work again I need to either restart VSCode, or open defer setting and switch it off and back on. After that I can run "Execute dbt SQL" again, but only until I try to run "Compiled dbt Preview". And looks like there is some kind of caching for "Compiled dbt Preview", because it breaks everything only when I can see a window "Compiling DBT model..." and it happens not for every click. 

Another observation is about target/compiled directory. Steps 3 or 4 writes compiled SQL code to the corresponding files here and use appropriate production dataset. But "Compiled dbt Preview" rewrites it with a "dev" dataset. Looks like "Execute dbt SQL" uses it somehow and there is some kind of caching or smth. 

Right now "Compiled dbt Preview" feature is not working for us at all.

**Comment by mdesmet** [comment]
This has been resolved in current version. Please verify and reopen if not working.

**Comment by arnaud-lecoules** [comment]
Hi, I still have the issue with 0.64.5. Can you reopen the issue ? 

**Setup**: 

- Power User for dbt 0.64.5 (darwin-arm64), VS Code
- dbt-core 1.12.0, dbt-snowflake 1.11.3, Python 3.12
- dbt.dbtIntegration: core
- Defer config set from the Actions panel (local mode):
```
"dbt.deferConfigPerProject": {
  "path/to/project": {
    "deferToProduction": true,
    "manifestPathForDeferral": "/abs/path/to/project/target-prod",
    "manifestPathType": "local",
    "favorState": false
  }
}
```
- `target-prod/manifest.json` is produced by `dbt parse --target PROD --target-path target-prod`.

**Steps to reproduce**:

1. Open model child.sql, which has {{ ref('parent') }}. parent exists in prod, not in the dev schema.
2. Build dbt Model → works. The command is invoked with --defer --state <target-prod> and parent resolves to its prod relation.
3. Execute dbt SQL (query preview, Cmd+Enter) on the same file → fails with Object '<DB>.<DEV_SCHEMA>.<DEV_ALIAS_OF_PARENT>' does not exist or not authorized. The compiled SQL points at the dev schema; deferral was not applied.
4. Compiled dbt Preview shows the same dev relation.

----------------------------------

Here I copy the diagnostic made by Claude (tell me if it only adds noise).

### Where it breaks

The preview path in the core integration spins up a fresh Python bridge for each query and does:

```
project = DbtProject(..., defer_to_prod=True, manifest_path=<target-prod>/manifest.json, ...)
project.init_project()
project.compile_sql(<query>)   # then execute_sql
```

In `dist/altimate_python_packages/dbt_core_integration.py`, `init_project()` → `create_parser()` loads the manifest through `ManifestLoader.load()`, but the deferred manifest is only merged in `safe_parse_project()`:

```python
def safe_parse_project(self) -> None:
    ...
    self.parse_project()
    self.write_manifest_artifact()
    if self.defer_to_prod:
        writable_manifest = WritableManifest.read_and_check_versions(self.defer_to_prod_manifest_path)
        manifest = Manifest.from_writable_manifest(writable_manifest)
        self.dbt.merge_from_artifact(other=manifest)
```

Because the preview process never calls `safe_parse_project()`, no node gets a `defer_relation`, and dbt's `RuntimeRefResolver.create_relation` falls back to the current target for every ref, whatever `args.defer` says.

I confirmed this by driving the shipped bridge directly (Snowflake `get_relation` stubbed to "relation absent"):

| Flow | `ref('parent')` compiles to |
|---|---|
| `DbtProject(defer_to_prod=True, manifest_path=...)` → `init_project()` → `safe_parse_project()` → `compile_sql` | `DB.prod_schema.parent` |
| same constructor → `init_project()` → `compile_sql` (what `executeSQL` does) | `DB.dev_schema.parent` |
| `set_defer_config(True, ...)` → `init_config()` → `safe_parse_project()` → `compile_sql` | `DB.prod_schema.parent` |

### Suggested fix

Either call the deferred-manifest merge in the per-query bridge after `init_project()` in `executeSQL` / `compileQuery` (core integration), or move the `merge_from_artifact` block from `safe_parse_project()` into `create_parser()` so every manifest load applies it. The second option also covers `Compiled dbt Preview` and any future caller.

Happy to test a pre-release build.

**Comment by ralphstodomingo** [comment]
_Automated triage note — posted from @ralphstodomingo's account by our extension triage / E2E bot._

@arnaud-lecoules thanks — your diagnosis is right, and the repro was precise enough to automate. We ran it against the **published** 0.64.5 and 0.64.6 builds with a matching setup: dbt-core 1.12.0, dbt-snowflake 1.11.3, Python 3.12, `dbt.dbtIntegration: core`, local-mode defer config with `favorState: false`, the state manifest produced by `dbt parse --target PROD --target-path target-prod`, `parent` built only in the PROD schema, `child.sql` = `select * from {{ ref('parent') }}`. The check reads what Snowflake actually received (query history), not the extension's own logs.

| journey | 0.64.5 | 0.64.6 |
|---|---|---|
| Execute dbt SQL on `child.sql`, fresh session (your step 3) | `<db>.<DEV>.parent` → `Object … does not exist or not authorized` | `<db>.<PROD>.parent` → succeeds ✅ |
| first Compiled dbt Preview | PROD | PROD |
| Execute dbt SQL again, after a preview | DEV → fails | DEV → fails ❌ |
| a second Compiled dbt Preview (your step 4) | DEV | DEV ❌ |
| Execute dbt SQL after a manifest rebuild (a model file saved) | — | PROD → succeeds ✅ |

So **0.64.6 fixes your step 3** — dbt-integration 0.3.13 compiles play-button queries on the fully parsed project (#2080) — but **not the whole report**. The Compiled dbt Preview handler refreshes the project configuration first, and that refresh re-initializes the project without re-applying the deferral manifest: the same `init_project()`-without-`safe_parse_project()` gap you described, reached from a second path. After a preview, the next Execute dbt SQL and every later preview resolve to the dev schema until the manifest is rebuilt. The first preview only looks right because its compile is queued ahead of the re-initialization.

We've filed the remainder internally; the fix goes in the direction of your second option (the deferral merge has to survive every manifest load) and will ship in a follow-up release.

**Workaround on 0.64.6 until then:** after opening a compiled preview, save any model file — that triggers a manifest rebuild, and Execute dbt SQL defers correctly again. For Compiled dbt Preview itself, only the first preview after a rebuild is correct.

Keeping this issue open until the follow-up ships.
